tests/test_typing.py

```diff
@@ -0,0 +1,70 @@
+# Copyright 2022 D-Wave Systems Inc.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+import itertools
+import os
+import re
+import unittest
+
+import numpy as np
+
+try:
+    import mypy
+    import mypy.api
+except ImportError:
+    mypy = None
+
+
+@unittest.skipUnless(mypy, "mypy not installed")
+@unittest.skipUnless(tuple(map(int, np.__version__.split('.')[:2])) >= (1, 20), "need numpy>=1.20")
+class TestTyping(unittest.TestCase):
+    TEST_DIR = os.path.join(os.path.dirname(__file__), 'typing')
+    CACHE_DIR = os.path.join(TEST_DIR, '.mypy_cache')
+
+    def test_mypy(self):
+        stdout, stderr, exit_code = mypy.api.run([
+            '--cache-dir', self.CACHE_DIR,
+            '--follow-imports', 'silent',
+            '--show-absolute-path',
+            self.TEST_DIR])
+
```

Comment:
Perhaps replace this python unittest with a CI job that just calls `mypy`?

Reply:
The problem is checking the failure cases. MyPy doesn't really have a very specific exclude syntax.